rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +68,237 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
+            self.notify_user(event=decoded_events[-1], counterparty=CPT_UNISWAP_V3)
+            return
+
+        # Check last two events and if they are related to the swap, label them as such
+        for event in last_events:
+            if (
+                event.balance.amount == asset_normalized_value(amount=amount_sent, asset=crypto_asset) and  # noqa: E501
+                event.event_type == HistoryEventType.SPEND
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.counterparty = CPT_UNISWAP_V3
+                event.notes = f'Swap {decoded_events[-1].balance.amount} {crypto_asset.symbol} in uniswap-v3'  # noqa: E501
+            elif (
+                event.balance.amount == asset_normalized_value(amount=amount_received, asset=crypto_asset) and  # noqa: E501
+                event.event_type == HistoryEventType.RECEIVE
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = CPT_UNISWAP_V3
+                event.notes = f'Receive {decoded_events[-1].balance.amount} {crypto_asset.symbol} as a result of a uniswap-v3 swap'  # noqa: E501
+
+    # --- Routers methods ---
+
+    def _decode_eth_to_token_swap(self, decoded_events: list[HistoryBaseEntry]) -> SwapData:
+        send_eth_event = decoded_events[1]
+        refund_eth_event = decoded_events[2] if decoded_events[2].asset == A_ETH else None
+        receive_token_event = decoded_events[2] if refund_eth_event is None else decoded_events[3]
+
+        from_amount = send_eth_event.balance.amount
+
+        if refund_eth_event is not None:
+            from_amount -= refund_eth_event.balance.amount
+
+        return SwapData(
+            from_asset=A_ETH,
+            from_amount=from_amount,
+            to_asset=receive_token_event.asset,
+            to_amount=receive_token_event.balance.amount,
+        )
+
+    def _decode_token_to_eth_swap(self, decoded_events: list[HistoryBaseEntry]) -> SwapData:
+        from_asset: Optional[Asset] = None
+        from_amount = ZERO
+
+        for i in range(2, len(decoded_events)):
+            event = decoded_events[i]
+            if (
+                event.event_type == HistoryEventType.INFORMATIONAL and
+                event.event_subtype == HistoryEventSubType.APPROVE
+            ):
+                continue  # skip approval events
+
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND and
+                event.counterparty in {CPT_UNISWAP_V2, CPT_UNISWAP_V3}
+            ) is False:
+                break  # the swap is over
+
+            if from_asset is None:
+                from_asset = event.asset
+            elif from_asset != event.asset:
+                break  # the swap is over
```

Comment:
> If I understood correctly in each iteration it detects the asset being swap and in the next event that is not informational it breaks it

Yes

> But why doesn't it finish in the if statement before this one

Yes, it should. I added this `break` just in case something from what we expect. Perhaps I'll change it so that if something unexpected happens, the decoding fails and the user is notified